src/arm.rs

```diff
@@ -128,18 +128,129 @@ extern "C" {
 
 // FIXME: The `*4` and `*8` variants should be defined as aliases.
 
-#[cfg_attr(not(test), no_mangle)]
-pub unsafe extern "aapcs" fn __aeabi_memcpy(dest: *mut u8, src: *const u8, n: usize) {
-    memcpy(dest, src, n);
-}
-#[cfg_attr(not(test), no_mangle)]
-pub unsafe extern "aapcs" fn __aeabi_memcpy4(dest: *mut u8, src: *const u8, n: usize) {
-    memcpy(dest, src, n);
-}
-#[cfg_attr(not(test), no_mangle)]
-pub unsafe extern "aapcs" fn __aeabi_memcpy8(dest: *mut u8, src: *const u8, n: usize) {
-    memcpy(dest, src, n);
-}
+// Assembly optimized memclr{4,8} and memset{4,8}
+//
+// Original source: https://github.com/bobbl/libaeabi-cortexm0/blob/15e10736afed6f947fba4b5740274fec4f2c95ca/memset.S
+// Author: Jörg Mische <bobbl@gmx.de>
+// License: ICS
+//
+// Copyright (c) 2013 Jörg Mische <bobbl@gmx.de>
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
+// OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+global_asm!(r"
+@ fn __aeabi_memclr4(r0: *mut u8, r1: usize)
+@ r0 is 4-byte aligned
+    .global __aeabi_memclr4
+
+@ fn __aeabi_memclr8(r0: *mut u8, r1: usize)
+@ r0 is 8-byte aligned
+    .global __aeabi_memclr8
+
+__aeabi_memclr4:
+__aeabi_memclr8:
+    @ r2 = r2 ^ r2 = 0
+    eors r2, r2
+
+    @ tail call __aeabi_memset4(r0, r1, 0)
+
+@ fn __aeabi_memset4(r0: *mut u8, r1: usize, r2: i32)
+@ r0 is 4-byte aligned
+    .global __aeabi_memset4
+
+@ fn __aeabi_memset8(r0: *mut u8, r1: usize, r2: i32)
+@ r0 is 8-byte aligned
+    .global __aeabi_memset8
+
+__aeabi_memset4:
+__aeabi_memset8:
+    @ branch to '2' if r1 < 4
+    subs r1, #4
+    blo  2f
+
+    @ copy r2 lower byte into all the other bytes of r2
+    @ e.g. r2 = 0xdead_beef -> r2 = 0xefef_efef
+    lsls r2, r2, #24
+    lsrs r3, r2, #8
+    orrs r2, r3
+    lsrs r3, r2, #16
+    orrs r2, r3
+
+1:  @ word-wise copy loop
+    str  r2, [r0]
+    adds r0, #4
+    subs r1, #4
+    bhs  1b
```

Comment:
Saw your RFC on IRC earlier, I think you can make this a little tighter for cores that support postfix operation. I wrote this ( https://gist.github.com/ppannuto/672328eb8184abdb9559 ) a little while back for a tight init loop, comparing:

```
str r2, [r0] (2 cycle)
adds r0, #4 (1 cycle)
subs r1, #4 (1 cycle)
bhs 1b (~1 cycle)
```
-> 5 cycles / loop

with stmia:
```
WriteZeroToBssStart:
  ldr r0, =_sbss
  ldr r1, =_ebss
  movs r2, #0
  b WriteZeroToBssEnterLoop
WriteZeroToBssLoop:
  stmia r0!, {r2}  <-- 2 cycles
WriteZeroToBssEnterLoop:
  cmp r0, r1  <-- 1 cycle
  bcc WriteZeroToBssLoop  <-- 1 cycle
```
-> 4 cycles / loop

Using cycle counts from 3.3.1 in http://infocenter.arm.com/help/topic/com.arm.doc.ddi0439b/DDI0439B_cortex_m4_r0p0_trm.pdf rather than measured, but I think this should save a cycle